### PR TITLE
RouteProgress#voiceInstructions property doesn't become null

### PR DESCRIPTION
**Pull request by VysotskiVadim**
### Description
The`RouteProgress#voiceInstructions` should not become null just like the `RouteProgress.bannerInstructions`. See #5017

I was trying to implement the feature using TDD. But I had problems with this. When I added `voiceInstructionsFollowRouteProgress` test to the `MapboxTripSessionTest` I realised a few things.

I don't trust the added tests. I wasn't sure that everything works when I saw a green color. Why? I mocked so many implementation details so that the code that I run in tests deviate from what our users run in their app. Green tests don't make me confident that the feature works in the app.

I can't refactor. I want to write a test, then write shitty implementation, then refactor it and make the implementation beautiful. The current approach forces me to setup many implementation details using `mockk`. When I refactor and change implementation details tests are breading and I have to rewrite them. I don't trust the test I've just rewritten. I have think a lot to avoid errors when I'm rewriting tests. I lost the main advantage that TDD gives me: easy refactoring.

In this PR I added a new test class `MapboxTripSessionNewTest` where I wrote tests as I usually do it. I considered `MapboxTripSession` as a unit and mocked only dependencies that are passed in constructor. I didn't consider static utils functions like `getRouteProgressFrom` as an external dependency, in my implementation it's considered as an implementation detail that can be changed without breaking the tests.

I used fake instead of a mock for `MapboxNativeNavigator`. The idea is to have reusable fake with different implemented scenario. It should behave very similar to a real one. So that when you use fake in tests you're confident that the code works with real navigator too. When the real navigator changes behaviour you just change one fake and run tests to see if other code still works. 
### Changelog

```
<changelog>
The `RouteProgress.voiceInstructions` property always keeps the last value and don't become `null`.
</changelog>
```

### CODE REVIEW

**Comment by Zayankovsky** [comment]
I don't agree that we shouldn't mock dependencies of the class. The idea behind unit tests is to test the logic in a single unit (class) only, so that we don't have to update the test, when other units change. That's why we have to isolate the class from its dependencies with the help of mocks. If you want to test the class with all its dependencies like in real scenarios, that's what integration tests are for. Also using fake implementations instead of mocks isn't really different, but mocks are much more concise and flexible.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt [hunk 1]
```diff
@@ -772,6 +774,28 @@ class MapboxTripSessionTest {
...
+
+        navigatorObserverImplSlot.captured.onStatus(navigationStatusOrigin, navigationStatus)
+
+        every { routeProgress.voiceInstructions } returns null
+
+        navigatorObserverImplSlot.captured.onStatus(navigationStatusOrigin, navigationStatus)
+
+        verify(exactly = 1) { voiceInstructionsObserver.onNewVoiceInstructions(any()) }
+        verify(exactly = 2) { routeProgressObserver.onRouteProgressChanged(any()) }
+    }
```
Comment (VysotskiVadim) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:797]: This is the test I don't trust. I leave it just for demonstration. Going to remove after feedback.

**Comment by VysotskiVadim** [comment]
@Zayankovsky , thank you for the feedback ❤️ 

> I don't agree that we shouldn't mock dependencies of the class

I agree with you. I provided test doubles for almost all dependencies.

> The idea behind unit tests is to test logic in a single unit (class) only, so that we don't have to update the test, when other units change.

In my understanding unit != class, it's up to developer to decide what is unit. I find [this article inspirational](https://blog.cleancoder.com/uncle-bob/2017/10/03/TestContravariance.html).

If your unit is more than one class it doesn't mean that you should update tests when some of the classes changes. When I tested implementation details using mocks I thought that every change of implementation leads to the changes of tests. But it works differently with the proposed style. You need to update tests only if you change behaviour or public API. If you change implementation details you just run exiting tests to see if the code works. Tests should be updated only if the behaviour of a feature changes.

In general I think that the classical TDD style is more tolerant to changes than mockist one. _Terms are taken from the [article](https://martinfowler.com/articles/mocksArentStubs.html), search for "Classical and Mockist Testing"._

> If you want to test the class with all its dependencies like in real scenarios, that's what integration tests are for.

I can test it there. But don't you want to get real scenarios in you fast and reliable unit tests for a low price? The closer scenario to a real one the most likely your tests find a bug.

> Also using fake implementations instead of mocks isn't really different, but mocks are much more concise and flexible.

I agree. Mocks are more flexible in stubbing than regular stubs. More power means more responsibility. Every time I used this additional flexibility I regretted using it during refactoring or other changes. I haven't seen any flexible feature from a mock framework that doesn't lead to a bad design. Have you seen?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/FakeNativeNavigator.kt [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+ * [FakeNativeNavigator] doesn't cover all scenarios.
+ * If you find an empty method instead of behavior emulation feel free to implement it.
+ * Don't hesitate to write tests for this test double.
+ *
+ * Some behavior can't be easily emulated.
+ * It's hard to parse a route and emulate location to trigger the status updates.
+ * Use [FakeNativeNavigator] as a stub for status update emulation.
+ * Call [updateStatus] method to trigger all observers.
+ */
+class FakeNativeNavigator private constructor() : MapboxNativeNavigator {
```
Comment (kmadsen) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/FakeNativeNavigator.kt:47]:
```suggestion
class TestNativeNavigator private constructor() : MapboxNativeNavigator {
```

Typically these classes are "Test*" classes. Referencing [TestCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-coroutine-dispatcher/index.html] [TestLifecycleOwner](https://developer.android.com/reference/kotlin/androidx/lifecycle/testing/TestLifecycleOwner),

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt (hunk 1 above)
Comment (kmadsen) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:797]: what do you mean you don't trust it?

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:797]: I mean the test doesn't give me confidence that feature works as expected. I know the logic which maps `NavigationStatus` to `RouteProgress` is mocked. The only real part that test verifies is 
image

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:797]: But maybe confidence isn't the main problem for me.

The main problem is I don't know how to work using TDD when I mock so many things. I want to state in the test the feature requirements "RouteProgress.voiceInstruction shouldn't be null if possible. When native navigator returns null in a status update, fallback to previous value but don't trigger the new voice instruction event". All I can do in this test is "When mapper returns null voice instruction trigger route progress event but don't trigger new voice instruction event".

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            createNavigationStatus(voiceInstruction = createVoiceInstruction("second"))
+        )
+        // assert
+        val announcementsFromVoiceInstructionObserver = voiceInstructionsObserver.records
+            .map { it.announcement() }
+        assertEquals(listOf("first", "second"), announcementsFromVoiceInstructionObserver)
+        val announcementsFromRouteProgress = routeProgressObserver.records
+            .map { it.voiceInstructions?.announcement() }
+        assertEquals(listOf("first", "first", "second"), announcementsFromRouteProgress)
+    }
```
Comment (VysotskiVadim) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:75]: The test states the same requirement as the feature description: _"RouteProgress.voiceInstruction shouldn't be null if possible. When native navigator returns null in a status update, fallback to previous value but don't trigger the new voice instruction event"_
Isn't it great. I was able to write a test that describe what I want and only then I started to think about implementation. Isn't it how TDD should work?

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:75]: Why was I able to write a test that states the same as the feature description? Classical TDD style considers tests as another client for a class. I.e. tests use a tested class just like other clients: access only public API and provide dependencies via a constructor, no interceptions somewhere in the middle. We write feature descriptions from the client's point of view as well.

That's why it's so easy to work using TDD with the classical TDD style. You write a test which is the same as the requirement. Then you think about implementation. Then you refactor without breaking and rewriting the tests.

**Comment by Zayankovsky** [comment]
I can also add that using real dependencies instead of mocks leads to the same logic being tested several times. So if we introduce a flow somewhere, we will have more failed tests and that will make it harder to find the bug.

**Comment by Zayankovsky** [comment]
Very often tests highlight problems with the way the code is structured. If we struggle to mock an external dependency, then we probably shouldn't have made it external in the first place. Also all external dependencies should be injected, e.g. via a constructor or a setter method. This is usually a better design than calling static methods, which also helps with testing.

**Comment by VysotskiVadim** [comment]
> Very often tests highlight problems with the way the code is structured. If we struggle to mock an external dependency, then we probably shouldn't have made it external in the first place. Also all external dependencies should be injected, e.g. via a constructor or a setter method. This is usually a better design than calling static methods, which also helps with testing.

This is the thing. I'm not struggling to mock the mapper. I consider it's as an implementation detail. Not mocking mapper makes tests simpler because... well you just write less code. + by not mocking a mapper I make tests tolerant to changes in it.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 4]
```diff
@@ -140,6 +140,7 @@ internal class MapboxTripSession(
     private val fallbackVersionsObservers = CopyOnWriteArraySet<FallbackVersionsObserver>()
 
     private val bannerInstructionEvent = BannerInstructionEvent()
+    private var lastVoiceInstruction: VoiceInstructions? = null
```
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:143]: I guess it should be set back to null in some cases. When don't we want to fallback to the previous value? I guess after route update. @LukasPaczos , do you know any scenarios when we don't want to fallback to the previous voice instruction?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt (hunk 1 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:798]: @LukasPaczos , the mapping is already tested in the bigger `voiceInstructionsFallbacksToPreviousValue` test. Do you think it's worth to test it one more time in the `NavigatorMapperTest` ?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 4 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:143]: That's correct, similarly to the `bannerInstrucitons`.

**Comment by Zayankovsky** [comment]
> This is the thing. I'm not struggling to mock the mapper. I consider it's as an implementation detail. Not mocking mapper makes tests simpler because... well you just write less code. + by not mocking a mapper I make tests tolerant to changes in it.

We can't consider the mapper an implementation detail, since it was extracted to a separate class. It can now be reused in any other class in this module. And if we follow your strategy, then we will have the mapper logic tested in multiple unrelated tests. This isn't what we want, because a bug in the mapper logic will likely cause all these tests to fail, which will make it harder to find the root cause.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt (hunk 1 above)
Comment (Zayankovsky) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionTest.kt:798]: You wouldn't have this question if the mapping was mocked in TripSession tests :)

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+ * [TestNativeNavigator] doesn't cover all scenarios.
+ * If you find an empty method instead of behavior emulation feel free to implement it.
+ * Don't hesitate to write tests for this test double.
+ *
+ * Some behavior can't be easily emulated.
+ * It's hard to parse a route and emulate location to trigger the status updates.
+ * Use [TestNativeNavigator] as a stub for status update emulation.
+ * Call [updateStatus] method to trigger all observers.
+ */
+class TestNativeNavigator private constructor() : MapboxNativeNavigator {
```
Comment (Zayankovsky) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt:48]: This looks huge and it will likely become only bigger if we start reusing this in another tests. 
Looks like it will even require tests on its own. 
I don't see any advantages of using this double over using mocks, which are very flexible and require much less code to write.

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt:48]: Are you saying that stubs has no benefits overs mock?

Comment (Zayankovsky) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt:48]: I think they are. Can you name some advantages of stubs?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 6]
```diff
@@ -620,11 +622,17 @@ internal class MapboxTripSession(
...
-                voiceInstructionsObservers.forEach {
-                    it.onNewVoiceInstructions(voiceInstruction)
-                }
+            triggerVoiceInstructionEvent(progress)
+        }
+    }
+
+    private fun triggerVoiceInstructionEvent(progress: RouteProgress) {
+        val voiceInstructions = progress.voiceInstructions
+        if (voiceInstructions != null && lastVoiceInstruction != voiceInstructions) {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: I'm wondering if checking for equality here might actual lead to some edge case bugs.

For example, if there's a "turn right in 50 meters" instruction and after we take that turn the next one is also "turn right in 50 meters". In such a case, we wouldn't emit the second instruction and leave a driver without audio guidance.

I think that we shouldn't check for equality here and always trust NN that if they emit a non-null instruction then we should use it and emit it as well.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt:48]: I share @Zayankovsky's sentiment as well, as we've been discussing yesterday @VysotskiVadim. Did you have a chance to try and migrate the stub factories to mock factories? How did it go?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 6 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: this is a good point

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: wait a minute. They should have different `distanceAlongGeometry`, shouldn't they? @LukasPaczos

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt (hunk 5 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt:48]: https://github.com/mapbox/mapbox-navigation-android/pull/5203

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/infra/TestNativeNavigator.kt]: merged.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 6 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: In an unfortunate case, that could be equal as well.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: Oh wait, maybe not :thinking:

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:631]: yep, you're right, they could be equals. _distanceAlongGeometry is the distance to next maneuver in unit meters_

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt]: fixed

**Comment by VysotskiVadim** [comment]
> And if we follow your strategy, then we will have the mapper logic tested in multiple unrelated tests.

I agree with mentioned disadvantages. If you test behaviour user expects you won't be pointed to the line which causes unexpected behaviour of internal logic. You will just know which behaviour that used to work doesn't work anymore. 

The tests you see in PR is superior to proposed in terms of refactoring. You can change implementation without breaking the tests, because tests verify behaviour our users expect from SDK, not specific implementation.

To be honest I can hardly imagine how to work using TDD by the suggested approach. It seems I should know implementation in advance to write some tests or write tests after implementation.

**Review by LukasPaczos — APPROVED** [review]
I still prefer us using mocks instead of stubs for unit test dependencies as they are more flexible, but the proposed tests get the job done and the logic itself looks good to me, so let's merge and iterate on the test setup if we need to.

Thanks for explaining your thought process behind this @VysotskiVadim!

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 7]
```diff
@@ -620,11 +622,18 @@ internal class MapboxTripSession(
...
+
+    private fun triggerVoiceInstructionEvent(progress: RouteProgress?, status: NavigationStatus) {
+        if (progress == null) return
+        val voiceInstructions = progress.voiceInstructions
+        val navigatorTriggeredNewInstruction = status.voiceInstruction != null
+        if (voiceInstructions != null && navigatorTriggeredNewInstruction) {
+            voiceInstructionsObservers.forEach {
+                it.onNewVoiceInstructions(voiceInstructions)
             }
+            lastVoiceInstruction = progress.voiceInstructions
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:636]:
```suggestion
        ifNonNull(routeProgress.voiceIntruction, (status.voiceInstruction != null)) { upcomingVoiceInstructions, isTriggeredNewInstruction -> 
          if(isTriggeredNewInstruction) { 
             voiceInstructionsObservers.forEach {
                it.onNewVoiceInstructions(upcomingVoiceInstructions)
            }
            lastVoiceInstruction = upcomingVoiceInstructions
          }
        }
```

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/navigator/NavigatorMapperTest.kt [hunk 8]
```diff
@@ -422,7 +427,8 @@ class NavigatorMapperTest {
             navigationStatus,
             mockk(relaxed = true),
             mockk(relaxed = true),
-            0
+            0,
+            lastVoiceInstruction = mockk(relaxed = true)
```
Comment (RingerJK) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/navigator/NavigatorMapperTest.kt:431]: nit 
```suggestion
            lastVoiceInstruction = mockk(relaxed = true),
```

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+class StatusWithVoiceInstructionUpdateUtil {
+    companion object {
+
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_1 = 1.0
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_NULL = 1.5
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_2 = 2.0
+
+        fun triggerStatusUpdatesOnLocationUpdate(nativeNavigator: MapboxNativeNavigator) {
```
Comment (RingerJK) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:279]: nit: make `StatusWithVoiceInstructionUpdateUtil` object instead class and restrict companion object

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/navigator/NavigatorMapperTest.kt [hunk 10]
```diff
@@ -94,7 +94,8 @@ class NavigatorMapperTest {
             navigationStatus,
             remainingWaypoints = 1,
             bannerInstructions,
-            instructionIndex = 1
+            instructionIndex = 1,
+            lastVoiceInstruction = null
```
Comment (RingerJK) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/navigator/NavigatorMapperTest.kt:98]: nit 
```suggestion
            lastVoiceInstruction = null,
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt [hunk 11]
```diff
@@ -62,13 +62,15 @@ internal fun getRouteProgressFrom(
     status: NavigationStatus,
     remainingWaypoints: Int,
     bannerInstructions: BannerInstructions?,
-    instructionIndex: Int?
+    instructionIndex: Int?,
+    lastVoiceInstruction: VoiceInstructions?
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt:66]: nit 
```suggestion
    lastVoiceInstruction: VoiceInstructions?,
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt [hunk 12]
```diff
@@ -86,7 +88,8 @@ private fun NavigationStatus.getRouteProgress(
     route: DirectionsRoute?,
     remainingWaypoints: Int,
     bannerInstructions: BannerInstructions?,
-    instructionIndex: Int?
+    instructionIndex: Int?,
+    lastVoiceInstruction: VoiceInstructions?
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/NavigatorMapper.kt:92]: nit 
```suggestion
    lastVoiceInstruction: VoiceInstructions?,
```

**Review by RingerJK — COMMENTED** [review]
nits

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 7 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:636]: don't you mind if I leave the code as is? I'm not a big fan of syntax sugar if it doesn't reduce lines count. BTW if you're insisting let's discuss it 🙂

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt (hunk 9 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:279]: Can't explain why, but I like that `triggerStatusUpdatesOnLocationUpdate` becomes static function in the bytecode. But you find it very ugly, I'm ready to accept singleton in the bytecode and replace `companion` by `object`. 🙂

Comment (Zayankovsky) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:279]: It doesn't become a static function in the bytecode, unless you add a `JvmStatic` annotation. 
It is actually a singleton `StatusWithVoiceInstructionUpdateUtil.Companion`. 
So basically there is no difference, but with `object` it is simpler. 
Also I think `triggerStatusUpdatesOnLocationUpdate` can just be a private function in test class or file.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 7 above)
Comment (Zayankovsky) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:636]: I also like the current version more. For me a simple `if` check looks cleaner than these custom `ifNonNull` functions.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt (hunk 9 above)
Comment (RingerJK) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:279]: `StatusWithVoiceInstructionUpdateUtil.Companion` is actually a static class in bytecode. There're 2 classes, and we can restrict it to only one

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt:279]: Ok, it's not how it works. I will fix

Comment (VysotskiVadim) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt]: fixed, thank you guys 👍
